File: src/core/execution/register_stack_engine.rs
#[derive(Clone, Copy, Debug)]
pub struct StackWindow {
    pub register_base: u64,
    pub size_of_frame: u64,

    pub count_input_registers: u64,
    pub count_local_registers: u64,
    pub count_output_registers: u64,
    pub count_rotating_registers: u64,
}

#[derive(Clone, Debug)]
pub struct RegisterStackEngine {
    stack_windows: Vec<StackWindow>
}

impl RegisterStackEngine {
    pub fn new() -> RegisterStackEngine {
        RegisterStackEngine { stack_windows: Vec::new() }
    }

    pub fn current_frame_mut(&mut self) -> &mut StackWindow {
        let index = self.stack_windows.len() - 1;

        return &mut self.stack_windows[index];
    }

    pub fn current_frame(&self) -> &StackWindow {
        let index = self.stack_windows.len() - 1;

        return &self.stack_windows[index];
    }

    //Creates a new frame after a function call
    pub fn new_frame(&mut self, input_registers: u64) {
        let current_frame = self.current_frame();

        self.stack_windows.push(StackWindow { 
            register_base: (current_frame.register_base + current_frame.size_of_frame) - input_registers, 
            size_of_frame: input_registers, 
            count_input_registers: input_registers, 
            count_local_registers: 0, 
            count_output_registers: 0, 
            count_rotating_registers: 0 
        })
    }

    //Allocates more registers if necessary
    pub fn allocate(&mut self, local_registers: u64, output_registers: u64) {
        let current_frame = self.current_frame_mut();

        current_frame.count_local_registers = local_registers;
        current_frame.count_output_registers = output_registers;

        current_frame.size_of_frame = current_frame.count_input_registers + current_frame.count_local_registers + current_frame.count_output_registers
    }
}

File: Rosalia.Core/Execution/RegisterStackEngine.cs
using System.Collections.Concurrent;

namespace Rosalia.Core.Execution;

public class StackWindow {
    public ulong RegisterBase;
    public ulong SizeOfFrame;

    public ulong CountInputRegisters;
    public ulong CountLocalRegisters;
    public ulong CountOutputRegisters;
    public ulong CountRotatingRegisters;
}

public class RegisterStackEngine {
    public readonly ConcurrentStack<StackWindow> StackWindows;

    public RegisterStackEngine() {
        this.StackWindows = new ConcurrentStack<StackWindow>();
    }

    public StackWindow CurrentFrame() {
        this.StackWindows.TryPeek(out StackWindow window);

        return window;
    }

    public void NewFrame(ulong inputRegisters) {
        StackWindow currentFrame = this.CurrentFrame();

        this.StackWindows.Push(new StackWindow {
            RegisterBase = (currentFrame.RegisterBase + currentFrame.SizeOfFrame) - inputRegisters,
            SizeOfFrame = inputRegisters,
            CountInputRegisters = inputRegisters,
            CountLocalRegisters = 0,
            CountOutputRegisters = 0,
            CountRotatingRegisters = 0
        });
    }

    public void Allocate(ulong localRegisters, ulong outputRegisters, ulong rotatingRegisters) {
        StackWindow currentFrame = this.CurrentFrame();

        currentFrame.CountLocalRegisters = localRegisters;
        currentFrame.CountOutputRegisters = outputRegisters;
        currentFrame.CountRotatingRegisters = rotatingRegisters;

        currentFrame.SizeOfFrame = currentFrame.CountInputRegisters + currentFrame.CountOutputRegisters + currentFrame.CountLocalRegisters + currentFrame.CountRotatingRegisters;
    }
}
